### PR TITLE
MultiError: Rename priv types, added support for errors.{As,Is}; tests.

### PR DESCRIPTION
**Pull request by bwplotka**
cc @pracucci 

Reason: We then don't need to cast MultiErr to type and access errors:  https://github.com/thanos-io/thanos/blob/master/pkg/compact/compact.go#L518

### CODE REVIEW

**Review by pracucci — APPROVED** [review]
Thanks Bartek! Should cover Thanos use cases.

**Comment by bwplotka** [comment]
````
+ go-fuzz-build -libfuzzer -func FuzzParseMetric -o promql-parse-metric.a ./promql
/go/src/github.com/prometheus/prometheus/tsdb/errors/errors.go:99:16: Is not declared by package errors
/go/src/github.com/prometheus/prometheus/tsdb/errors/errors.go:86:16: As not declared by package errors
typechecking of ./promql failed
```

Most likely we need to fix fuzzing Go version first.

**Comment by brancz** [comment]
Generally happy with this :+1:

**Review by kakkoyun — COMMENTED** [review]
Nice 👍

File: tsdb/errors/errors.go
Comment (codesome) [inline tsdb/errors/errors.go:97]: Should we also require all errors in `target` to match with `es`? Or is it just the _type_ of the error that should match?

Comment (bwplotka) [reply tsdb/errors/errors.go:97]: As written in docs.. it's up to us:

```
// Is reports whether any error in err's chain matches target.
//
// The chain consists of err itself followed by the sequence of errors obtained by
// repeatedly calling Unwrap.
//
// An error is considered to match a target if it is equal to that target or if
// it implements a method Is(error) bool such that Is(target) returns true.
//
// An error type might provide an Is method so it can be treated as equivalent
// to an existing error. For example, if MyError defines
//
//	func (m MyError) Is(target error) bool { return target == os.ErrExist }
//
// then Is(MyError{}, os.ErrExist) returns true. See syscall.Errno.Is for
// an example in the standard library.
```

Will add comment (:

**Comment by bwplotka** [comment]
This is blocked on lack of https://hub.docker.com/r/fuzzitdev/golang/tags version for modern Go versions. Does anyone have an idea who to contact about this? :hugs: Or replacement image? Tried with normal golang image and it's lacking some compiler directive. Wonder if it's just image with this: https://go.googlesource.com/proposal/+/master/design/draft-fuzzing.md implemented?

**Comment by roidelapluie** [comment]
Hi @bwplotka , we have looked at this during our bug scrub, and were wondering where would you use this in the Prometheus code base? Multierr would become a lot more complicated.

**Comment by bwplotka** [comment]
Technically any user of Prometheus/TSDB code can't use `errors.Is` `errors.As` against our APIs which can lead to surprises.

**Comment by roidelapluie** [comment]
> Technically any user of Prometheus/TSDB code can't use `errors.Is` `errors.As` against our APIs which can lead to surprises.

How is that semantically correct anyway, since there are multiple errors, how can they be tailed down to just one?

File: tsdb/errors/errors.go
Comment (bwplotka) [inline tsdb/errors/errors.go:82]:
```suggestion
// As finds the first error in multiError slice of error chains that matches target, and if so, sets
```

File: tsdb/errors/errors.go
Comment (bwplotka) [inline tsdb/errors/errors.go:89]: I assumed initially that making type priv means we user can't As(multiErr), but that's not true (to verify)

< Offline 1:1 with @codesome >

File: tsdb/errors/errors.go
Comment (bwplotka) [reply tsdb/errors/errors.go:106]: You are right.

< Offline 1:1 with @codesome >

**Review by bwplotka — COMMENTED** [review]
Some comments during our 1:1, but overall we think it's not harmful (and only useful) to have this in Prom

**Comment by bwplotka** [comment]
Rebased, addressed comments, and extended tests. Added `Count` function as well.

**Comment by beorn7** [comment]
Note that I plan to cut the v2.24 RC in two days (2020-12-30). This needs to get merged in time to get in.

**Comment by bwplotka** [comment]
Yea, I did not have any explicit approve on this from any maintainer. Only "generally" looking good (: 

@brancz @roidelapluie @codesome ?

Also I was not pushing this too much as we needed dependency lightweight version of multi error in many other Go projects, so I ported it here: https://pkg.go.dev/github.com/efficientgo/tools/core

Planning to use it in Thanos etc instead, but happy for suggestions (:

**Comment by brancz** [comment]
If there are no objections from others, I'm happy with this.

lgtm :+1:

File: tsdb/errors/errors_test.go
Comment (roidelapluie) [inline tsdb/errors/errors_test.go:1]: How accurate is that?

Comment (bwplotka) [reply tsdb/errors/errors_test.go:1]: How that happened? 😱

**Review by bwplotka — COMMENTED** [review]
Thanks for review @roidelapluie 

I ported this package to separate minimal module so we could all use it thousands of other projects we do 🤗 

Wonder if we would be willing to use it instead of this in Promethues?

See: https://pkg.go.dev/github.com/efficientgo/tools/core@v0.0.0-20210129205121-421d0828c9a6/pkg/merrors 

Yes, it's kind of `common` but with (almost) no dependencies and core, focused functionalities only.

**Comment by roidelapluie** [comment]
I do not *feel* fine with importing efficientgo, for a few reasons.

The first one is that it is pretty new and has a single developer, which means that we are stuck with plan b's if there are bugs. Also it appears that it is not really used a lot by other projects.

It appears that that library does not do releases, which is a problem because our dependencies upgrades won't work on it.

**Comment by bwplotka** [comment]
Thanks @roidelapluie , valid concerns. Let me address those: 

> The first one is that it is pretty new and has a single developer, which means that we are stuck with plan b's if there are bugs. Also it appears that it is not really used a lot by other projects.

I am happy to give Prometheus Team write access, that's not a problem.

Upgradability: I don't think that matters here. Our dependencies upgrade will work, just nothing will be ever upgraded for efficientgo repo until we do a manual pin (which is what we what, since we control that module).

I would argue that maintaining the same library in thousands of places is a worse alternative. Anyway, not a blocker, happy to wait for bigger adoption to change your mind first. In meantime let's copy the same `merrors` in Prometheus. 

Will address comments on this PR.

**Comment by roidelapluie** [comment]
We have looked at this pull request during our bug scrub.

Commit 344c630857 introduced Errors.Is, so I think we can close this pull request.

If you need `As()` support, feel free to open a new PR.

Thank you for your contribution.

**Pull request closed without merging**
